### ISSUE TITLE
Add non-cyclic rule for external object construction

### ISSUE THREAD

**Issue by henrikt-ma** [modelica/ModelicaSpecification#2845]
This is a follow-up to #1907.  As I understand the intention of that discussion, we want to avoid entangled external object construction.  At the same time, I expect that we want more freedom than having to construct all external objects even before solving the initialization problem.

What I think we want is something along these lines (perhaps right after the sentence with _called exactly once before the first use of the object_):
> The arguments to the constructor must not – directly nor indirectly – depend on the external object being constructed.

Here's an example of a cyclic dependency that would be illegal:
```
model CyclicExternalObjectConstruction

  class EO
    extends ExternalObject;
    …
  end EO;

  function get
    input EO obj;
    output Real y;
    …
  end get;

  EO obj = EO(x);
  parameter Real x = get(obj);

end CyclicExternalObjectConstruction;
```

**Comment by henrikt-ma** [comment]
See also #1705.

**Comment by beutlich** [comment]
> 
> 
> Also closely related to #1705.

Please have in mind though #1705 is closed it is not yet resolved, It only was moved to #2037.

**Comment by henrikt-ma** [comment]
Another closely related issue is #2661.  In view of #2037, one could probably think of #2661 as drawing the line between using the `constructor` and the `reconstructor`.

**Comment by henrikt-ma** [comment]
> Please have in mind though #1705 is closed it is not yet resolved, It only was moved to #2037.

For the example given in #2037, wouldn't the present proposal provide a simpler solution?  On the other hand, doesn't #2037 have more compelling applications for calling external object constructors with non-parametric arguments?

**Comment by HansOlsson** [comment]
Agreement to add.
